packages/statemanager/src/cache/code.ts

```diff
@@ -0,0 +1,277 @@
+import { bytesToUnprefixedHex, equalsBytes } from '@ethereumjs/util'
+import debugDefault from 'debug'
+import { OrderedMap } from 'js-sdsl'
+import { LRUCache } from 'lru-cache'
+
+import { Cache } from './cache.js'
+import { CacheType } from './types.js'
+
+import type { CacheOpts } from './types.js'
+import type { Address } from '@ethereumjs/util'
+
+const { debug: createDebugLogger } = debugDefault
+
+/**
+ * Represents a cached code element.
+ */
+type CodeCacheElement = {
+  code: Uint8Array | undefined
+}
+
+export class CodeCache extends Cache {
+  _lruCache: LRUCache<string, CodeCacheElement> | undefined
+  _orderedMapCache: OrderedMap<string, CodeCacheElement> | undefined
+
+  /**
+   * Diff cache collecting the state of the cache
+   * at the beginning of checkpoint height
+   * (respectively: before a first modification)
+   *
+   * If the whole cache element is undefined (in contrast
+   * to the code), the element didn't exist in the cache
+   * before.
+   */
+  _diffCache: Map<string, CodeCacheElement | undefined>[] = []
+
+  constructor(opts: CacheOpts) {
+    super()
+    if (opts.type === CacheType.LRU) {
+      this._lruCache = new LRUCache({
+        max: opts.size,
+        updateAgeOnGet: true,
+      })
+    } else {
+      this._orderedMapCache = new OrderedMap()
+    }
+
+    this._diffCache.push(new Map<string, CodeCacheElement | undefined>())
+    this._debug = createDebugLogger('statemanager:cache:code')
+  }
+
+  _saveCachePreState(
+    cacheKeyHex: string,
+    currentPutCode?: Uint8Array | undefined,
+    currentCode?: Uint8Array | undefined
+  ) {
+    const it = this._diffCache[this._checkpoints].get(cacheKeyHex)
+    if (it === undefined) {
+      let oldElem: CodeCacheElement | undefined
+      if (this._lruCache) {
+        oldElem = this._lruCache.get(cacheKeyHex)
+      } else {
+        oldElem = this._orderedMapCache!.getElementByKey(cacheKeyHex)
+      }
+
+      let val
+      if (currentCode !== undefined && equalsBytes(currentCode, new Uint8Array())) {
+        val = undefined
+      } else {
+        val = currentPutCode
+      }
+      this._diffCache[this._checkpoints].set(cacheKeyHex, oldElem ?? { code: val })
+    }
+  }
+
+  /**
+   * Puts code into the cache under its hash.
+   * @param codeHash - Hash of the code.
+   * @param code - Bytecode or undefined if code doesn't exist.
+   */
```

Comment:
TODO: docs. I am not entirely sure what `currentCode` does as opposed to `code`.

Reply:
Fixed. Let me know if I can further clarify.